### PR TITLE
TreeTraversal added

### PR DESCRIPTION
**Pull request by ram123-debug**
BSTree traversal like inOrder, preOrder and postOrder functions are added with O(n) time and O(n) space complexity

[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm? 
- [ ] Documentation change? 

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md). 
- [x] This pull request is all my own work -- I have not plagiarized. 
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Data-Structures/Tree/TreeTraversal/Test.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+
```
Comment (raklaptudirm) [inline Data-Structures/Tree/TreeTraversal/Test.js:1]: We use the `jest` package for testing. Make the necessary changes.

File: Data-Structures/Tree/TreeTraversal/TreeTraversal.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+//O(n) Time and O(n) Space Complexity
+
```
Comment (raklaptudirm) [inline Data-Structures/Tree/TreeTraversal/TreeTraversal.js:2]: Format your code with `standard.js`.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]

**Comment by ram123-debug** [comment]
Not able meet your requirement , need help

**Comment by raklaptudirm** [comment]
Please elaborate what you need help on.

**Comment by ram123-debug** [comment]
Not able to pass the the Continuous Integration / Code style and tests (pull_request) . Please fix it

**Comment by raklaptudirm** [comment]
You must install the [`standard.js`](https://standardjs.com/) package and run the `standard --fix .` command on the files.

**Comment by ram123-debug** [comment]
standard fix is done

File: Data-Structures/Tree/TreeTraversal.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+function postOrderTraverse (tree, array) {
+  if (tree !== null) {
+    postOrderTraverse(tree.left, array)
+    postOrderTraverse(tree.right, array)
+    array.push(tree.value)
+  }
+}
+
+exports.inOrderTraverse = inOrderTraverse
```
Comment (raklaptudirm) [inline Data-Structures/Tree/TreeTraversal.js:28]: This repository uses Ecma Script Modules to handle packages, not Common JS.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]

File: Data-Structures/Tree/TreeTraversal.js (hunk 3 above)
Comment (raklaptudirm) [reply Data-Structures/Tree/TreeTraversal.js]: Format the code after every change.

File: Data-Structures/Tree/TreeTraversal.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/* eslint no-use-before-define: 0 */ // --> OFF
```
Comment (raklaptudirm) [inline Data-Structures/Tree/TreeTraversal.js:1]: Why is this exemption necessary?

Comment (ram123-debug) [reply Data-Structures/Tree/TreeTraversal.js]: used it for debugging purpose now removed

File: Data-Structures/Tree/TreeTraversal.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+
```
Comment (raklaptudirm) [inline Data-Structures/Tree/TreeTraversal.js:1]: Remove this empty line.

File: Data-Structures/Tree/TreeTraversal.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+function inOrderTraverse (tree, array) {
+  if (tree !== null) {
+    inOrderTraverse(tree.left, array)
+    array.push(tree.value)
+    inOrderTraverse(tree.right, array)
+  }
+  return array
+}
+function preOrderTraverse (tree, array) {
```
Comment (raklaptudirm) [inline Data-Structures/Tree/TreeTraversal.js:12]: Add an empty line before this.

**Review by raklaptudirm — DISMISSED** [review]
Some formatting stuff.

File: Data-Structures/Tree/TreeTraversal.js (hunk 6 above)
Comment (ram123-debug) [reply Data-Structures/Tree/TreeTraversal.js:12]: done

**Review by raklaptudirm — DISMISSED** [review]

File: Data-Structures/Tree/TreeTraversal.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+function inOrderTraverse (tree, array) {
+  if (tree !== null) {
+    inOrderTraverse(tree.left, array)
+    array.push(tree.value)
+    inOrderTraverse(tree.right, array)
+  }
+  return array
+}
+
+function preOrderTraverse (tree, array) {
```
Comment (CarlosZoft) [inline Data-Structures/Tree/TreeTraversal.js:12]: this function should return an array, right ?

File: Data-Structures/Tree/TreeTraversal.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  if (tree !== null) {
+    array.push(tree.value)
+    preOrderTraverse(tree.left, array)
+    preOrderTraverse(tree.right, array)
+  }
+}
+
+function postOrderTraverse (tree, array) {
```
Comment (CarlosZoft) [inline Data-Structures/Tree/TreeTraversal.js:20]: this function should return an array, right ?

File: Data-Structures/Tree/TreeTraversal.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  if (tree !== null) {
+    postOrderTraverse(tree.left, array)
+    postOrderTraverse(tree.right, array)
+    array.push(tree.value)
+  }
+}
+
+export { preOrderTraverse }
+export { inOrderTraverse }
```
Comment (CarlosZoft) [inline Data-Structures/Tree/TreeTraversal.js:29]: would be better :
```
  export { preOrderTraverse, inOrderTraverse, postOrderTraverse }
```
what do you think?

**Review by CarlosZoft — CHANGES_REQUESTED** [review]

**Review by CarlosZoft — APPROVED** [review]
LGTM 🔥

File: Data-Structures/Tree/TreeTraversal.js (hunk 9 above)
Comment (ram123-debug) [reply Data-Structures/Tree/TreeTraversal.js]: Yeah you are right

**Review by raklaptudirm — APPROVED** [review]

**Review by appgurueu — CHANGES_REQUESTED** [review]
Please make it apparent that these are *binary* tree traversals rather than *general* tree traversals; I also don't like pushing the results to an array, I'd rather use a visitor or generator pattern here as to not incur the O(n) space complexity.

Furthermore, please consider integrating this (making it a bunch of methods) with our existing [binary (search) tree implementation](https://the-algorithms.com/de/algorithm/binary-search-tree?lang=javascript).

Also please add tests & some more comments.

**Pull request closed without merging**